### PR TITLE
Bug/en 3192 fix bug when a wrong/unused fork requested header is received

### PR DESCRIPTION
**Pull request by SebastianMarian**
This implementation should fix a situation when all peers would been replaced in their headerNonceHash pool a good/used header in their blockchain construction, with a wrong/unused header on which they didn't construct, but which came after a late broadcast from a valid proposer.

### CODE REVIEW

File: process/sync/metablock.go [hunk 1]
```diff
@@ -376,6 +377,20 @@ func (boot *MetaBootstrap) doJobOnSyncBlockFail(hdr *block.MetaBlock, err error)
 			log.Info(errNotCritical.Error())
 		}
 	}
+
+	// The below section of code fixed a situation when all peers would been replaced in their headerNonceHash pool a
```
Comment (AdoAdoAdo) [inline process/sync/metablock.go]: would *have* replaced

File: process/sync/shardblock.go [hunk 2]
```diff
@@ -487,6 +488,20 @@ func (boot *ShardBootstrap) doJobOnSyncBlockFail(hdr *block.Header, err error) {
 			log.Info(errNotCritical.Error())
 		}
 	}
+
+	// The below section of code fixed a situation when all peers would been replaced in their headerNonceHash pool a
```
Comment (AdoAdoAdo) [inline process/sync/shardblock.go]: would have replaced

Comment (sasurobert) [reply process/sync/shardblock.go]: +1

File: process/sync/metablock.go (hunk 1 above)
Comment (sasurobert) [reply process/sync/metablock.go]: +1

File: process/sync/metablock.go [hunk 3]
```diff
@@ -376,6 +377,20 @@ func (boot *MetaBootstrap) doJobOnSyncBlockFail(hdr *block.MetaBlock, err error)
 			log.Info(errNotCritical.Error())
 		}
 	}
+
```
Comment (AdoAdoAdo) [inline process/sync/metablock.go]: I can see there is a lot of duplicated code between these two files, if there is common functionality, we should extract into a common interface and not have the code duplicated.

For now the hotfix can go in like this, but please open a refactor issue for this.

File: process/sync/metablock.go (hunk 1 above)
Comment (SebastianMarian) [reply process/sync/metablock.go]: Done

File: process/sync/shardblock.go (hunk 2 above)
Comment (SebastianMarian) [reply process/sync/shardblock.go]: Done

File: process/sync/metablock.go (hunk 3 above)
Comment (AdoAdoAdo) [inline process/sync/metablock.go]: Just took a look on the function, and at line 370 "isForkDetected" - this variable name seems misleading, why is a fork detected when we have a timeout error? this seems not logical, so maybe use a better name for this.

Comment (SebastianMarian) [reply process/sync/metablock.go]: Ok. Added task EN-3195 in jira for this

Comment (SebastianMarian) [reply process/sync/metablock.go]: Agree. Renamed the variable

**Pull request merged**